File: go/stalin-sort.go
package stalinSort

func SortInts(arr []int) []int {
	sorted := make([]int, 0)

	for i := 0; i < len(arr); i++ {
		if i == 0 {
			sorted = append(sorted, arr[i])
		} else if arr[i] > sorted[len(sorted)-1] {
			sorted = append(sorted, arr[i])
		}
	}

	return sorted
}

File: C/stalin-sort.c
#include "stalin-sort.h"
#include <string.h> // for memcpy, memset

/* Sorts array by non-decreasing
 * Interface is similar to qsort
 *
 * arr - array begin
 * count - number of elements in array
 * size - sizeof of 1 element
 * cmp - pointer on comparison function
 * 
 * Size of sorted array is written to count
 * Freed space is filled by zeros
 */
void stalin_sort( void *arr, size_t *count, size_t size,
           		  int (*cmp)(const void *, const void *) )
{
	if (*count <= 1)
		return;

	void *last = arr; // last sorted element
	void *current = arr + size; // current considered element

	while (--*count)
	{
		if ( cmp(current, last) >= 0 && current > (last += size))
			memcpy(last, current, size);
		current += size;
	}

	last += size;
	memset(last, 0, current - last); // Writing zeros to freed space in array. It is convenient when sorting chars or c-strs

	*count = (last - arr) / size;
}